docs/proposals/improved-schema-for-services.md

```diff
@@ -0,0 +1,409 @@
+
+# Improved schema for operator backed services
+
+This proposal is about improving the experience for a user of the operator backed services. Currently a user doesn't know exaustively which fields are available in a CR as we depend on optional metadata present in CRDs, to improve this we are considering using the metadata available from the cluster about the CRDs.
+Getting that metadata from the cluster is challenging because a normal cluster user ( plain vanilla ) doesn't have access to that metadata so we are gonna follow a 3 tiered approach
+
+- check if the user has access to the CRD, if so then use that.
+- get the `swagger.json` from the openapi endpoint provided by kubernetes. a sample can be found here https://gist.github.com/girishramnani/f29949d5cb8c6547102776437e05ac19
+- finally if we cannot get the openapi data as well then use the `ClusterServiceVersion` to get the parameters.
+
+This change would affect multiple service commands and the changes are described briefly below -
+- `odo catalog describe service` should include the metadata fields with description so the users can provide then when doing `odo service create`
+- `odo catalog list service` shouldn't change much other then listing CRDs per operator
+- `odo service create` should take flags and dynamically fill the CRD structs - metadata would be used for validation
+- `odo service create --from-file` would be used by adapter team or whoever wants to provide the whole CRD themselves as a file. this feature is already present and working.
+
+# Implementation plan
+
+Note - Below I would mention `cluster` manytimes and that means both openshift and kuberenetes. If there is something specific to one or the other then I would mention it.
+
+Below is the step wise plan for implementing this feature - 
+
+## Getting the metadata
+
+### Admin Access/Access to CRD API path
+First we would consider the scenerio where the user has admin access or privileges to access the CRD API to the cluster.
```

Comment:
We would implement all 3 scenarios. As it’s implemented in webconsole

Reply:
That seems overkill to me. If there's one approach that would work for admin as well as non-admin users, we should stick with that and improve upon that instead of trying to cover multiple scenarios. So far, it seems like swagger.json information is that thing which works for all types of users.

Reply:
I understand but this proposal is designed towards having similar experience in odo as web console.

swagger approach also has its dis-advantages and if the feedback after the swagger implementation is that it’s enough then we can drop it. But for the proposal as it is trying to provide feature parity with web console. I have kept it.

Reply:
Also after the swagger approach is implemented. It’s gonna be comparatively straightforward to implement the admin approach

Reply:
Makes sense.